internal/cryptutil/tls.go

```diff
@@ -0,0 +1,105 @@
+package cryptutil
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+
+	"github.com/caddyserver/certmagic"
+	"github.com/go-acme/lego/v3/challenge/tlsalpn01"
+)
+
+// effectiveCA is used to test Let's Encrypt without hitting production
+// usage limits.
+var effectiveCA = certmagic.LetsEncryptStagingCA
+
+// NewAutocert automatically retrieves public certificates from the free
+// certificate authority Let's Encrypt using HTTP-01 and TLS-ALPN-01 challenges.
+// To complete the challenges, the server must be accessible from the internet
+// by port 80 or 443 .
+//
+// https://letsencrypt.org/docs/challenge-types/#http-01-challenge
+// https://letsencrypt.org/docs/challenge-types/#tls-alpn-01
+func NewAutocert(tlsConfig *tls.Config, hostnames []string, path string) (*tls.Config, func(h http.Handler) http.Handler, error) {
+	certmagic.DefaultACME.Agreed = true
+	certmagic.DefaultACME.CA = effectiveCA
+	cm := certmagic.NewDefault()
+
+	tlsConfig = newTLSConfigIfEmpty(tlsConfig)
+	// add existing certs to the cache, and staple OCSP
+	for _, cert := range tlsConfig.Certificates {
+		if err := cm.CacheUnmanagedTLSCertificate(cert, nil); err != nil {
```

Comment:
autocert is called last so that any existing cert can be added to autocert's cache.